libp2p/security/insecure/transport.py

```diff
@@ -21,12 +44,29 @@ class InsecureTransport(BaseSecureTransport):
         for an inbound connection (i.e. we are not the initiator)
         :return: secure connection object (that implements secure_conn_interface)
         """
-        return InsecureSession(self, conn, ID(b""))
+        session = InsecureSession(self, conn, ID(b""))
+        await session.run_handshake()
+        return session
 
     async def secure_outbound(self, conn: IRawConnection, peer_id: ID) -> ISecureConn:
         """
         Secure the connection, either locally or by communicating with opposing node via conn,
         for an inbound connection (i.e. we are the initiator)
         :return: secure connection object (that implements secure_conn_interface)
         """
-        return InsecureSession(self, conn, peer_id)
+        session = InsecureSession(self, conn, peer_id)
+        await session.run_handshake()
+        # TODO: Check if `remote_public_key is not None`. If so, check if `session.remote_peer`
+        received_peer_id = session.get_remote_peer()
+        if received_peer_id != peer_id:
+            raise SecurityUpgradeFailure(
```

Comment:
What do you think about raising exception like `HandshakeFailure` here and catch it upstream and then raise `UpgradeFailure` there(in `TransportUpgrader.upgrade_security`)?
`HandshakeFailure` here -> catch by `SecurityMultistream` and raise `MultiselectFailure` -> catch by `TransportUpgrader` and raise `UpgradeFailure`.